src/adapter/src/rbac.rs

```diff
@@ -418,6 +448,417 @@ fn ownership_err(
+                BTreeSet::from([(
+                    item.name().qualifiers.clone().into(),
+                    AclMode::USAGE,
+                    role_id,
+                )])
+            }
+            ObjectId::Cluster(_) | ObjectId::Database(_) | ObjectId::Role(_) => BTreeSet::new(),
+        },
+        Plan::CreateDatabase(_)
+        | Plan::CreateRole(_)
+        | Plan::CreateCluster(_)
+        | Plan::DiscardTemp
+        | Plan::DiscardAll
+        | Plan::EmptyQuery
+        | Plan::ShowAllVariables
+        | Plan::ShowVariable(_)
+        | Plan::SetVariable(_)
+        | Plan::ResetVariable(_)
+        | Plan::StartTransaction(_)
+        | Plan::CommitTransaction(_)
+        | Plan::AbortTransaction(_)
+        | Plan::CopyRows(_)
+        | Plan::AlterNoop(_)
+        | Plan::AlterSystemSet(_)
+        | Plan::AlterSystemReset(_)
+        | Plan::AlterSystemResetAll(_)
+        | Plan::AlterRole(_)
+        | Plan::Declare(_)
+        | Plan::Fetch(_)
+        | Plan::Close(_)
+        | Plan::Prepare(_)
+        | Plan::Execute(_)
+        | Plan::Deallocate(_)
+        | Plan::Raise(_)
+        | Plan::RotateKeys(_)
+        | Plan::GrantRole(_)
+        | Plan::RevokeRole(_) => BTreeSet::new(),
+    }
+}
+
+fn generate_required_source_privileges(
+    catalog: &impl SessionCatalog,
+    plan: &CreateSourcePlan,
+    depends_on: &Vec<GlobalId>,
+    role_id: RoleId,
+) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
+    let mut privileges = BTreeSet::from([(
+        plan.name.qualifiers.clone().into(),
+        AclMode::CREATE,
+        role_id,
+    )]);
+    if let SourceSinkClusterConfig::Existing { id } = &plan.cluster_config {
+        privileges.insert((id.into(), AclMode::CREATE, role_id));
+    }
+    privileges.extend(generate_item_usage_privileges(catalog, depends_on, role_id));
+    privileges
+}
+
+/// Generates all the privileges required to execute a read that includes the objects in `ids`.
+///
+/// Not only do we need to validate that `role_id` has read privileges on all relations in `ids`,
+/// but if any object is a view or materialized view then we need to validate that the owner of
+/// that view has all of the privileges required to execute the query within the view.
+///
+/// For more details see: <https://www.postgresql.org/docs/15/rules-privileges.html>
+fn generate_read_privileges(
+    catalog: &impl SessionCatalog,
+    ids: impl Iterator<Item = GlobalId>,
+    role_id: RoleId,
+) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
+    generate_read_privileges_inner(catalog, ids, role_id, &mut BTreeSet::new())
+}
+
+fn generate_read_privileges_inner(
+    catalog: &impl SessionCatalog,
+    ids: impl Iterator<Item = GlobalId>,
+    role_id: RoleId,
+    seen: &mut BTreeSet<(GlobalId, RoleId)>,
+) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
+    let mut privileges = BTreeSet::new();
+    for id in ids {
+        if !seen.contains(&(id, role_id)) {
+            seen.insert((id, role_id));
```

Comment:
```suggestion
        if seen.insert((id, role_id)) {
```

Unsure if this is clearer or not? Ignore if you think it's fine as is.

Reply:
I think your approach is a bit cleaner, I just updated it.